File: javascript/3.Longest-Substring-Without-Repeating-Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  let length = 0;
  let left = 0;
  const set = new Set();

  for (let right = 0; right < s.length; right++) {
    while (set.has(s[right])) {
      set.delete(s[left]);
      left += 1;
    }

    length = Math.max(length, right - left + 1);
    set.add(s[right]);
  }

  return length;
};

File: python/3.Longest-Substring-Without-Repeating-Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        length = 0
        left = 0
        seen = set()

        for right in range(len(s)):
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            
            length = max(length, right - left + 1)
            seen.add(s[right])

        return length